### ISSUE TITLE
Swap file conflicts can result in ATTENTION message not being shown

### ISSUE THREAD

**Issue by benburrill** [vim/vim#2685]
### EDIT: Most (if not all) of this is not a bug with vim
As jamessan commented, `//` should be used at the end of directories.  Problems can still occur if `*.swo` files do get created, but this is rare and I am no longer sure about the value of fixing this.

---

I have `'directory'` set to a directory other than `.` (see Background).  Because of this, files with identical basenames create `*.swo` (etc) files in my `'directory'`.  However, vim does not check `*.swo` files when determining whether to display the ATTENTION message.

To reproduce:
 - Create `a/file` and `b/file`
 - Open both concurrently with `vim -N -u NONE --cmd 'set directory=~/.vim/swap'` (first a, then b).  Edit them a bit for kicks.
 - Open `a/file` in a new instance of vim (same options).  This displays the ATTENTION message as desired, although it does have the somewhat undesirable effect of prompting for which swap file you want to use for recovery, despite the swap files referring to different paths.
 - Open `b/file` in a new instance of vim (same options).  This does not display the ATTENTION message!
- Even if you close `a/file` (removing the `*.swp`), vim still doesn't check for the `*.swo` upon opening `b/file`.  I feel this might be problematic even when `'directory'` is set to `.`.

---

**Background** :  
My distro recently decided to push `directory=~/.vim/swap` on its users due to CVE-2017-1000382.  This "sane default" has caused several problems for me, mostly not vim's fault.  
Although I am not enthused by my distro's decision, I currently do not plan to reset it in my vimrc.  In theory, I like the setting -- if `'directory'` worked like `'undodir'`, I would already be using something similar. (EDIT: I guess I didn't rtfm very closely)
Due to CVE-2017-1000382 (which honestly seems meh to me), many vim users will probably end up (forced or not) switching to a similar configuration, so accommodating it well seems wise.

**Comment by jamessan** [comment]
The correct way to put all swap files into a single directory is to add a `//` to the end of the directory name, as mentioned in [`:help 'directory'`](https://vimhelp.appspot.com/options.txt.html#'directory'):

> For Unix and Win32, if a directory ends in two path separators "//"
> or "\\", the swap file name will be built from the complete path to
> the file with all path separators substituted to percent '%' signs.
> This will ensure file name uniqueness in the preserve directory.

**Comment by tonymec** [comment]
@benburrill : I don't know which distro you're using, but mine also pushes "insane defaults" on its unsuspecting users via a "system vimrc" at /etc/vimrc.

My answer is to compile my own, leaving Vim defaults in place, so it will look for a system vimrc at $VIM/vimrc (i.e. /usr/share/vim/vimrc) instead, and not find it, so that my vimrc is sourced with Bram's good old 'nocompatible' defaults the way it is described in the help.

The distro's build is a 7.4.326, strangely enough without patch 7.4.208 "Mercurial picks up files that are not distributed (empty)" which might be irrelevant. What troubles me more is that that source is dated 14 June 2014, which is starting to get, shall we say, a little long in the teeth, so I have absolutely no compunctions about building a more "modern" Vim for my own use.

**Comment by benburrill** [comment]
@jamessan Ah, I see.  Thanks!  I'll report that to my distro.  I still think this could be a problem even when `'directory'` is at its (true) default value, but it's definitely a lot less likely, and maybe not worth the hassle.  The easiest way to reproduce the problem is by choosing "edit anyway" in the recover screen.

**Comment by chrisbra** [comment]
Yes, it can happen even if swap files are located in its original location. For example my Vim workdir contains a whole bunch of `.swo` `.swp` `:swn`... .files, because I test a lot when fixing thigs and sometimes Vim crashes even. As soon as the `.swp` file is missing, Vim won't check for other older swap files so does not present the ATTENTION dialoge anymore. I have even come used to it but this might be a problem for users. I think we even have a ticket about it somewhere.
